src/smolagents/logger.py

```diff
@@ -0,0 +1,274 @@
+import json
+from dataclasses import dataclass
+from enum import IntEnum
+from typing import Any, Dict, List, Optional
+
+from rich.console import Console
+from rich.rule import Rule
+from rich.syntax import Syntax
+
+from smolagents.models import MessageRole
+from smolagents.utils import AgentError, make_json_serializable
+
+
+console = Console()
+
+
+class AgentStepLog:
+    pass
+
+    def dict(self):
+        raise NotImplementedError
+
+    def to_messages(self, **kwargs) -> List[Dict[str, Any]]:
+        raise NotImplementedError
+
+
+@dataclass
+class Message:
+    role: MessageRole
+    content: str
+
+    def dict(self):
+        return {"role": self.role, "content": self.content}
+
+
+@dataclass
+class ToolCall:
+    name: str
+    arguments: Any
+    id: str
+
+    def dict(self) -> str:
+        return {
+            "id": self.id,
+            "type": "function",
+            "function": {
+                "name": self.name,
+                "arguments": make_json_serializable(self.arguments),
+            },
+        }
+
+
+@dataclass
+class ActionStep(AgentStepLog):
```

Comment:
You could probably make a lot of the logic here way more succinct by using a library like `monty.json`. It has a `as_dict()` method which does this recursive casting to dictionaries. Actually, `dataclass` has a [similar method](https://stackoverflow.com/a/72605423) which may be even easier. But perhaps that's for another PR.

Reply:
We try to avoid too many external imports in libraries to keep them lightweight, and I don't think the default asdict from dataclass supports non primitives, but I'll simplify the code with it where I can, good idea!